### PR TITLE
Disallow _ in node name

### LINKED ISSUE #1915: Acceptable characters in naming of yaml.
**Operating environment/Installation (Hass.io/Docker/pip/etc.):**

**ESP (ESP32/ESP8266, Board/Sonoff):**

**ESPHome version (latest production, beta, dev branch)**

**Affected component:**

**Description of problem:**

**Problem-relevant YAML-configuration entries:**
```yaml
PASTE YAML FILE HERE
My problem ended up being with _ in my yaml names. It uses the name of the yaml for dhcp lookup and my lynksys router didn't handle those request quiet right. My router would randomly change the node ip. Digging deeper into my dhcp client tables some clients would have other clienst names and not the name with _. It's a fairly new router and once I reserved ips for the mac addresses with affected names I'm having no issues and reboots are much faster. I'd recommened removing the _ as accetpable.
```

**Logs (if applicable):**

```
PASTE DEBUG LOG HERE

```

**Additional information and things you've tried:**

### PR DESCRIPTION
**Pull request by pkuehne**
# What does this implement/fix? 

This removes underscore as a valid character in the name of a node. This is because the name of the node is used as the hostname and underscores are not valid in hostnames.

This may now stop previously working yaml files from passing validation.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Configuration change (this will require users to update their yaml configuration files to keep working)

**Related issue or feature (if applicable):** fixes esphome/issues#1915

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1072
  
# Test Environment

- [ ] ESP32
- [ ] ESP8266
- [X] Windows
- [ ] Mac OS
- [ ] Linux

# Explain your changes

As per the Linux [manpage](https://man7.org/linux/man-pages/man7/hostname.7.html) the only valid characters for a hostname are lowercase characters, numbers and hyphen. I've also updated the suggestion feature in the wizard to replace any underscores with hyphens, like it does with spaces, so it can't generate invalid configs.

I've split the name validation into two parts: `valid_name` and `valid_node_name`, as the former was also used to validate names of services in the `api` component and there was no good reason why they would need to change.

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by OttoWinter** [comment]
Hmm this would be a hard no from my side. The problem is that changing a node name is a very difficult process (see the ota page, you need to manually set the use_address and it's a multiple step process), and all affected users would have to go through that process. That simply doesn't outweigh the benefit of it not working with some routers.

Even printing a warning message would be a bit too much, unless we give users a simple button to click to correct this issue.

Was allowing underscores kn the hostname wrong? Yes, but iirc it was because something in HA's mqtt discovery (which was the primary transport in the beginning) didn't support hyphens.

However I think there could be another solution: For all users with existing YAML files we can assume their setup at least works partially (or 99%). What if we just remove the underscore in all setup wizards (esphome name.yaml wizard and dashboard wizard). Then new users are guided to the correct decision and we don't have a very big breaking change.

**Comment by glmnet** [comment]
Since I have 10+ more node with `_` in the hostname I disagree this too.

A warning mechanism is not very well implemented in ESPHome yet so not really a good option until that is done (I expect an error-like system where those can be linked to config / path / text ranges in the code.
Also a mechanism to disable warnings so you can do clean compilations

**Comment by pkuehne** [comment]
Yeah, most of my yamls have underscores in them as well. I'll revert the changes to the validation and just keep the wizard changes.

**Comment by glmnet** [comment]
Having said that, I do see some warning printed by ESPHome (e.g. Time zone not detected, using some FLASH SPI as GPIO etc), another one will not make much damage until proper implemented, look for LOGGER.LOG_WARN or something like that.

**Comment by pkuehne** [comment]
@OttoWinter @glmnet I've adjusted the code so it now prints a warning, but allows underscores in the name when running `compile` etc. The wizard still suggests you change underscores to hyphens, so that going forward it generates well-formed hostnames.

**Comment by OttoWinter** [comment]
From my side I would prefer not to print a warning at all for existing users.

With a warning most users will notice and try to find a way to correct it. We will get lots of reports from people and have to explain each time the complicated process to change a node's name, thus not only using up our resources but also making users unhappy.

And to what end? If a user compiles a node and notices the warning, that means they've used that device before, so at the very least their setup already works. In those cases that would just be extra work for little benefit.

If we print a warning, we should expect users to want to fix the warning, so in that case we should have at the very least some actionable information or button that does the process for them, especially when the benefit of such action is limited.

File: esphome/wizard.py [hunk 1]
```diff
@@ -195,11 +195,11 @@ def wizard(path):
...
                     "red",
                     f'Oh noes, "{name}" isn\'t a valid name. Names can only '
-                    f"include numbers, lower-case letters, underscores and "
-                    f"hyphens.",
+                    f"include numbers, lower-case letters and hyphens. ",
                 )
             )
-            name = strip_accents(name).lower().replace(" ", "_")
+            name = strip_accents(name).lower().replace(" ", "-")
+            name = strip_accents(name).lower().replace("_", "-")
```
Comment (OttoWinter) [inline esphome/wizard.py:202]: The same should be done for the setup wizard in the dashboard (see index.html in the dashboard templates directory)

Comment (pkuehne) [reply esphome/wizard.py:202]: ah thank you, I've not used the wizard dashboard before

**Comment by pkuehne** [comment]
My only counterpoint would be that not all yamls are "existing users". I haven't used the wizard after the initial setup and now copy+paste my yamls. The first one didn't need an underscore, the others all have it. If I'd been warned, I could have avoided using them.

On the other hand as you said, the benefit is limited and most people won't experience any adverse effects. So I will revert the changes and instead update the documentation to add a sentence on why underscores should be avoided when naming nodes.

**Comment by glmnet** [comment]
> And to what end?

Maybe I misunderstood the real issue here, what do you mean with:

> That simply doesn't outweigh the benefit of it not working with some routers.

What does not work?

**Comment by pkuehne** [comment]
> What does not work?

Have a look at the issue that sparked this PR: esphome/issues#1915

It seems that some routers get confused if a node has a hostname with an underscore, since underscores are not normally allowed in hostnames.

**Comment by glmnet** [comment]
Well, I'd rather read a warning than figure out my router changes my IP nodes addresses

**Comment by stewiem2000** [comment]
> The problem is that changing a node name is a very difficult process (see the ota page, you need to manually set the use_address and it's a multiple step process), and all affected users would have to go through that process.

Could a lot of that pain be relatively easily mitigated with a "try hyphen-based hostname first, retry with a underscore-based hostname" during the OTA phase?

**Comment by pkuehne** [comment]
@OttoWinter Could you take another look? I've updated the dashboard wizard as well.

**Comment by OttoWinter** [comment]
> > The problem is that changing a node name is a very difficult process (see the ota page, you need to manually set the use_address and it's a multiple step process), and all affected users would have to go through that process.
> 
> Could a lot of that pain be relatively easily mitigated with a "try hyphen-based hostname first, retry with a underscore-based hostname" during the OTA phase?

The problem is that this doesn't just affect OTA. If you change the name you also need to reconnect with HA core, all unique IDs (and thus entity ids) will change, etc etc.

**Review by OttoWinter — APPROVED** [review]
Thanks! :tada:

**Pull request merged**